x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -593,10 +571,16 @@ static Set<String> fieldNames(LogicalPlan parsed, Set<String> enrichPolicyMatchF
                 enrichRefs.removeIf(attr -> attr instanceof EmptyAttribute);
                 references.addAll(enrichRefs);
             } else if (p instanceof LookupJoin join) {
-                keepJoinReferences.addAll(join.config().matchFields());  // TODO: why is this empty
                 if (join.config().type() instanceof JoinTypes.UsingJoinType usingJoinType) {
                     keepJoinReferences.addAll(usingJoinType.columns());
                 }
+                if (keepCommandReferences.isEmpty()) {
+                    // No KEEP commands after the JOIN, so we need to mark this index for "*" field resolution
+                    wildcardJoinIndices.add(((UnresolvedRelation) join.right()).table().index());
+                } else {
+                    // Keep commands can reference the join columns with names that shadow aliases, so we block their removal
+                    keepJoinReferences.addAll(keepCommandReferences);
```

Comment:
Added both unit tests and csv-spec tests for multiple joins on the same lookup index, and it all works as expected. The rules for each index is that if there is a KEEP after that index, it uses fieldNames, otherwise it needs to use `*`, regardless of how many times that index is joined with or if any previous joins had a KEEP afterwards.